README.md
<div align="center">
  <a href="https://github.com/tomma5o/touchmyripple">
    <img src="./docs/images/double-tap.png" width="116px">
  </a>
  <br>
  <h1>TouchMyRipple</h1>
  <p>
    <em>touchMyRipple is a simple library that integrate ripple effect in your fantastic site!</em>
  </p>
  <p>
    <a href="https://github.com/tomma5o/touchmyripple/commits/master">
      <img src="https://img.shields.io/github/last-commit/tomma5o/touchmyripple.svg" />
    </a>
    <img src="https://img.shields.io/npm/dm/localeval.svg" />
  </p>
  <br>
  <a href="https://tomma5o.github.io/touchMyRipple/#line3"><b>Demo and Docs</b></a>
</div>

## Installation
 
[**Download Build**](https://raw.githubusercontent.com/tomma5o/touchMyRipple/master/lib/touchMyRipple.js)

```html
<script src="myDirectory/touchMyRipple.js"></script>
```
or
```javascript
npm install touchmyripple --save
```

## Basic Usage

`index.js`
```javascript
import tmripple from 'touchmyripple';

tmripple.init({
    color: '#bada55', // default is 'rgba(255, 255, 255, 0.4)'
    eventListener: 'touchstart' // default is 'click'
});

tmripple.attachToSelectors({
    selectors: '#foo .bar [type=button]',
    color: 'rgba(0, 0, 0, 0.4)',
    eventListener: 'mousedown'
});
```

## Methods

### **init(settings[Object])**

This method enables ripple effect to all the elements that have the attribute _data-animation="ripple"_.<br>
_All the following options are optional_


`settings[Object]`
```javascript
{
    // area is an option to make data-animation more specific
    area: 'class, id',

    // color...🤔 
    color: 'rgba, hex, hsla', 

    // pass the scrolling element if it's not window
    offsetEl: 'class, id',

    // this option accept an event listener 
    eventListener: 'event'
}
```

### **attachToSelectors(settings[Object])**

This method enables ripple effect to all the elements that match the class passed in 'selectors'

`settings[Object]`
```javascript
    {
        // selector of the element you want to attach the ripple ( is required )
        selectors: 'class, id',

        // color...🤔🤔🤔🤔🤔🤔
        color: 'rgba, hex, hsla',

        // pass the scrolling element if it's not window
        offsetEl: 'class, id',

        // this option accept an event listener 
        eventListener: 'event'
    }
```
## React Usage

`button.jsx`
```jsx
import withRipple from 'touchmyripple/react';

class Button extends Component {
  render() {
    return (
      <button {...this.props}>
        Hello <span>World</span>
      </button>
    );
  }
}

export default withRipple(Button, {
  color: "red",
  ignoreEls: { type: "secondary" }
});
```

`app.js`
```jsx
class App extends React.Component {
  render() {
    const settingObj = {
      eventName: "click", 
      color: "blue"
    }

    return  <Button tmripple={settingObj} />
    // or use it with default settings: <Button  />
  }
}
```

`settings[Object]`
```javascript
    {
        // default is "rgba(255,255,255,0.5)"
        color: 'rgba, hex, hsla',

        // this option accept an event listener for
        // differentiate smartphone event from desktop events
        // default is "click"
        eventListener: 'event',

        // ignore a specific element
        disabled: true

        // 🔺 this setting can be setted ONLY in the decorator function
        // you can create an object of key/value attributes to ignore
        ignoreEls: {class: "CTA_Button"}
    }
```

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [2.7.0] - 2018-07-29
### Added
- Global setting as a second parameter of the decorator function
- `disabled` option in the setting
- `ignoreEls` option in the setting

## [2.6.2] - 2018-07-25
### Fixed
- Props propagation inside a component

## [2.6.11] - 2018-07-25
### Fixed
- Changed requestAnimation frame to a normal setTimeOut with delay of 50ms, for timing problems

## [2.6.01] - 2018-07-25
### Added
- React HOC for add ripple in every element you want
- Helper function for creating a react folder when you run ```npm run build``` in local

### Changed
- File Tree
- Dist folder changed to lib

## [2.0.0] - 2018-07-23
### Added
- Gzipped Version

### Changed
- Changed event (only desktop) from ```Event.x``` to ```Event.clientX``` (more stable)

### Fixed
- Bug in event bubbling - now if we have a span inside a button the ripple start in the right element

## [1.6.6] - 2018-07-18

### Fixed
- Small refactor in the core

touchMyRipple.js
const touchMyRipple = () => {
    function ripple(els, rippleColor) {
        for (var i = 0; i < els.length; i += 1) {
            var currentBtn = els[i];

            currentBtn.addEventListener('touchstart', function (e) {
                var el = this.getBoundingClientRect(),
                    btnWidth = this.clientWidth,
                    rippleOffset = tmripple.settings.offsetEl,
                    headerHeight = rippleOffset ? rippleOffset.clientHeight : 0,
                    btnOffsetTop = el.top + headerHeight,
                    btnOffsetLeft = el.left,
                    posMouseX = e.changedTouches[0].pageX,
                    posMouseY = e.changedTouches[0].pageY + headerHeight,
                    rippleX = posMouseX - btnOffsetLeft,
                    rippleY = posMouseY - btnOffsetTop;

                var baseCSS = `position: absolute;
                               width: ${btnWidth * 2}px;
                               height: ${btnWidth * 2}px;
                               border-radius: 50%;
                               transition: transform 700ms, opacity 700ms;
                               transition-timing-function: cubic-bezier(0.250, 0.460, 0.450, 0.940);
                               background: ${rippleColor};
                               top: ${rippleY - btnWidth}px;
                               left: ${rippleX - btnWidth}px;
                               transform: scale(0);
                               pointer-events: none;`;

                // Prepare the dom
                var rippleEffect = document.createElement('span');
                rippleEffect.style.cssText = baseCSS;

                // Add some css for prevent errors
                this.style.overflow = 'hidden';

                if (window.getComputedStyle(this).position !== 'absolute' || window.getComputedStyle(this).position !== 'fixed') {

                    this.style.position = 'relative';
                }

                this.appendChild(rippleEffect);

                // start animation
                setTimeout(() => {
                    rippleEffect.style.cssText = `${baseCSS} transform: scale(1); opacity: 0;`;
                }, 5);

                setTimeout(() => {
                    rippleEffect.remove();
                }, 700);
            });
        }

    }
    function attachRippleToAttribute (area, rippleColor) {

        var attributeEl = document.querySelectorAll(`${area} [data-animation='ripple']`);

        if (attributeEl.length > 0) {
            ripple(attributeEl, rippleColor);
        } else {
            throw new Error('Selector/s not found');
        }
    }

    function attachRippleToSelectors(selectors, rippleColor) {
        if (selectors) {
            var selectorsEl = document.querySelectorAll(selectors);
        } else {
            throw new Error('You have to enter at least 1 selector');
        }


        if (selectorsEl.length <= 0) {

            console.warn('No element found');
        }

        ripple(selectorsEl, rippleColor);
    }

    var tmripple = {

        settings: {
            area: '',
            color: 'rgba(255, 255, 255, 0.4)',
            offsetEl: null,
            eventListener: 'click',
        },

        init(data) {
            try {
                this.settings.area = (data && data.area) ? data.area : '';
                this.settings.color = (data && data.color) ? data.color : this.settings.color;
                this.settings.offsetEl = (data && data.offsetEl) ? this.setOffsetEl(data.offsetEl) : null;
                this.settings.eventListener = (data && data.eventListener) ? data.eventListener : 'click';


                attachRippleToAttribute(this.settings.area, this.settings.color);
            } catch (e) {
                console.error(e.message);
                console.error(e);
            }
        },

        attachToSelectors(data) {
            try {
                var rippleColor = data.color || this.settings.color;

                attachRippleToSelectors(data.selectors, rippleColor);
            } catch (e) {
                console.error(e.message);
                console.error(e);
            }
        },

        setOffsetEl(el) {
            this.settings.offsetEl = document.querySelector(el);
        },
    };
    return tmripple;
};

module.exports.tmripple = touchMyRipple();
